Fix all Qt 5.15 deprecation warnings

c4rlo opened the pull request:
Fix all deprecation warnings when building against Qt 5.15. This is a first step towards being able to eventually upgrade to Qt 6 (see #7774 for discussion).

This now builds cleanly with `-DWITH_XC_ALL=ON -DWITH_GUI_TESTS=ON -DWITH_DEV_BUILD=ON`.

Since this is a sizeable change, I've done my best to structure this sensibly into a set of commits.

I've assumed here that the earliest Qt version we support is 5.9.5, as stated in INSTALL.md. I have adjusted the check in CMakeLists.txt accordingly (0a1e77656505ec79492d371e9e4c8a3f9f632e74). In some cases I've resorted to compile-time checks for the Qt version.

One particular annoyance is that  `QUrl::topLevelDomain()` has been [deprecated without direct replacement](https://bugreports.qt.io/browse/QTBUG-80308). I worked around that by accessing its functionality indirectly (9ed6095eb4c3e19d89783bae4ea89640ae5928b4).

## Testing strategy

Built with `-DWITH_XC_ALL=ON -DWITH_GUI_TESTS=ON -DWITH_DEV_BUILD=ON -DCMAKE_BUILD_TYPE=Debug` and ran all tests.

Did the same thing in an Ubuntu 18.04 container that has `libbotan-kpxc-2-dev` from [`ppa:phoerious/keepassxc`](https://launchpad.net/~phoerious/+archive/ubuntu/keepassxc) installed in it. This one did cause some test failures (in `TestCli::testClip()` and `TestCli::testInvalidDbFiles()`), but those also failed against the `2.7.4` tag, so I feel okay about them.

## Suggestion

In order to keep the code base deprecation-clean, it would be excellent if the automated tests (i.e. TeamCity) could be enhanced to build with `-DWITH_XC_ALL=ON -DWITH_GUI_TESTS=ON -DWITH_DEV_BUILD=ON -DCMAKE_BUILD_TYPE=Debug` and ideally against different Qt versions, including Qt 5.9.5 (assuming that we want to keep supporting it) and Qt 5.15 (the last Qt 5 version).

## Type of change

- ✅ Refactor (significant modification to existing code)

c4rlo committed:
rm QDateTime::toString(Qt::DefaultLocaleShortDate)

QDateTime::toString(Qt::DefaultLocaleShortDate) will go away in Qt 6;
replace it as recommended in the docs, introducing a Clock::toString()
helper function.

The replacement functionality we use already exists in Qt 5.9.5, our
minimum supported Qt version.

c4rlo committed:
{QString->Qt}::SkipEmptyParts

QString::SplitBehavior (of which SkipEmptyParts is an element) was
deprecated in Qt 5.15.

Its designated replacement, Qt::SplitBehavior, was only added in Qt
5.14. Hence we need to use some preprocessor magic to keep supporting
older Qt versions (we require at minimum Qt 5.9.5).

c4rlo committed:
Don't call deprecated QFlags(nullptr) ctor

It is deprecated as of Qt 5.15.

c4rlo committed:
Don't call deprecated QList::toSet, QSet::fromList

Both methods are deprecated since Qt 5.14.

The recommended replacement

  QSet(list.begin(), list.end())

is only available from Qt 5.14. Since we support Qt 5.9.5 or better,
provide our own 'Tools::asSet()' helper function for this purpose.

c4rlo committed:
QSet::{toList->values}

QSet::toList() is deprecated as of Qt 5.14.

Use the recommended replacement QSet::values() instead. It already
exists as of Qt 5.9.5, our minimum supported Qt version.

c4rlo committed:
rm QDateTime::toString(Qt::SystemLocaleShortDate)

Replace it with

  QLocale::system().toString(..., QLocale::ShortFormat)

as recommended in the official docs.

The replacement already exists in Qt 5.9.5, our earliest supported Qt
version.

c4rlo committed:
QHash::insertMulti -> QMultiHash::insert

QHash::insertMulti is deprecated as of Qt 5.15.

QMultiHash already exists in Qt 5.9.5, our earliest supported Qt
version.

c4rlo committed:
QProcess::startDetached: non-deprecated overload

The overload we were using before has been deprecated in Qt 5.15. To get
the same behaviour as before, we need to use QProcess::splitCommand(),
which was only introduced in Qt 5.15, hence we need to use an #if for
the Qt version.

c4rlo committed:
QPainter::{HighQuality->}Antialiasing

The "HighQuality" version is deprecated as of Qt 5.14.

Since it's not immediately clear at which point in Qt version history it
becomes preferable to use the non-"HighQuality" one, we only do so in Qt
5.14 onwards.

c4rlo committed:
Use QDateTime::startOfDay()

...instead of QDate(QDateTime) constructor, which is deprecated as of Qt
5.15.

Since QDateTime::startOfDay() is only available in Qt 5.14, we need to
use an #if.

c4rlo committed:
qSort() -> std::sort()

qSort() is deprecated.

c4rlo committed:
QPalette::{background->window}()

QPalette::background() is deprecated as of Qt 5.13.

Its replacement, QtPalette::window(), is already available in Qt 5.9.5,
our minimum supported Qt version.

c4rlo committed:
Use Qt::{Background,Foreground}Role

Instead of their deprecated aliases.

tenzap commented:
FYI, with 2.7.0 (+ c33995e0755bcbbe1b8e535da716a17b336a8277)  it is still possible to build with Qt 5.5 (but not below)

c4rlo commented:
> FYI, with 2.7.0 (+ [c33995e](https://github.com/keepassxreboot/keepassxc/commit/c33995e0755bcbbe1b8e535da716a17b336a8277)) it is still possible to build with Qt 5.5 (but not below)

Hmm, so what is the minimum Qt version the maintainers want to support? Currently, 5.9.5 is what's documented (in INSTALL.md).

c4rlo committed:
Use QRecursiveMutex

Instead of deprecated constructor QMutex(QMutex::RecursionMode), use
recommended replacement class QRecursiveMutex.

This deprecation was done only in the KDE branch of Qt, specifically in
https://invent.kde.org/qt/qt/qtbase/-/commit/6d5988831862f081404270bd7fa2f25bc836fdc9.
Since that commit is included in the Qt packaged for some Linux
distributions (e.g. Arch Linux:
https://archlinux.org/packages/extra/x86_64/qt5-base/), it seems worth
respecting this deprecation.

Since QRecursiveMutex was only introduced in Qt 5.14, whereas we need to
support Qt 5.9.5 or better, this involves some #ifdef magic.

c4rlo committed:
QProcess::{pid->processId}

QProcess::pid() has been deprecated since Qt 5.15. Recommended
replacement QProcess::processId() exists since Qt 5.3.

c4rlo committed:
endl -> Qt::endl, flush -> Qt::flush

Qt 5.15 deprecates usage of non-'Qt::'-prefixed 'endl' and 'flush' I/O
manipulators.

Since the 'Qt::' versions are only introduced in Qt 5.14, we backport
them for older Qt versions (since we support down to Qt 5.9.5).

c4rlo commented:
Ok, I think this is in a reasonable state now. Regarding the test failures:

## Ubuntu failure

From the [logs](https://ci.keepassxc.org/viewLog.html?buildId=211511&buildTypeId=KeePassXC_DebugClang&tab=buildLog), this is:

```
[12:58:26]
[Step 3/3] FAIL!  : TestKdbx4Format::testCustomData() Compared values are not the same
[12:58:26]
[Step 3/3]    Loc: [/opt/buildagent/work/c401303cba1b4098/tests/TestKdbx4.cpp(471)]
```

That line is:

```c++
    QCOMPARE(*clonedEntry->customData(), *entry->customData());
```

Now, `Entry::clone()` calls `CustomData::copyDataFrom()`, which calls `CustomData::updateLastModified()`, which can modify the custom data if we're just on the cusp of a new second; so I suspect that might be a pre-existing bug (likely affecting only the tests) that happens occasionally.

## Windows failure

Not much detail in the logs, just:

```
[13:01:48]
[Step 3/4] 1/3 Test #39: testgui ..........................***Failed   23.33 sec
```

Since I don't have a Windows build setup ready to go, I'd appreciate any hints or details.

droidmonkey commented:
You can ignore the gui failures, they are transient

c4rlo commented:
Cool. My latest update just rebased against `develop`; good to see that both failures have gone away. From my POV, this now seems good to merge.

The CustomData failure is as I suspected, I'll open a separate PR for that.

droidmonkey commented on src/CMakeLists.txt:
```diff
@@ -320,7 +320,7 @@ configure_file(config-keepassx.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config-keepas
 
 add_library(autotype STATIC ${autotype_SOURCES})
-target_link_libraries(autotype Qt5::Core Qt5::Widgets)
+target_link_libraries(autotype Qt5::Core Qt5::Network Qt5::Widgets)
```
Why was this added?

c4rlo replied:
I think this was because `autotype_SOURCES` includes `core/Tools.cpp` (even before my change), which now uses [`QNetworkCookie`](https://doc.qt.io/qt-5/qnetworkcookie.html), which requires the Qt Network module.

c4rlo replied:
Yup this is needed. If I remove it, I get errors like:

```
build] /home/carlo/src/keepassxc/qt-modernize/src/core/Tools.cpp:39:10: fatal error: QNetworkCookie: No such file or directory
[build]    39 | #include <QNetworkCookie>
[build]       |          ^~~~~~~~~~~~~~~~
```

droidmonkey commented on src/cli/Utils.h:
```diff
@@ -18,6 +18,8 @@
 #ifndef KEEPASSXC_UTILS_H
 #define KEEPASSXC_UTILS_H
 
+#include "core/Global.h"
+
```
Add this to the cpp instead

droidmonkey commented on src/core/Global.h:
```diff
@@ -42,6 +45,32 @@
+} // namespace Qt
+#endif
+
+template <class T> QSet<T> asSet(const QList<T>& a)
+{
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+    return QSet<T>(a.begin(), a.end());
+#else
+    return QSet<T>::fromList(a);
+#endif
+}
```
Arguably this should be in core/Tools.h|cpp

droidmonkey commented on src/core/Tools.cpp:
```diff
@@ -165,6 +168,12 @@ namespace Tools
         return QObject::tr("%1 minute(s)", nullptr, minutes).arg(minutes);
     }
 
+    QString toString(const QDateTime& dateTime)
+    {
+        static QLocale locale;
+        return locale.toString(dateTime, QLocale::ShortFormat);
+    }
```
This function name is too generic for it's purpose. Either add this to core/clock.h|cpp or name it something like datetimeToString(...)

droidmonkey commented on src/core/Tools.cpp:
```diff
@@ -296,6 +305,54 @@ namespace Tools
+            static const auto dummy = QByteArrayLiteral("x");
+            QNetworkCookie cookie(dummy, dummy);
+            cookie.setDomain(parentDomain);
+            if (!QNetworkCookieJar{}.setCookiesFromUrl(QList{cookie}, url)) {
```
Clever!

c4rlo commented:
I have addressed all your comments (by amending my commits, rather than pushing new ones) and rebased against `develop`.

Locally, all tests (including GUI tests) continue to succeed for me on my Arch Linux machine with Qt 5.15. In an Ubuntu 18.04 container it also succeeds apart from the `TestCli::testClip()` and `TestCli::testInvalidDbFiles()` failures, but those also fail against `2.7.4`.

Likewise, the current TeamCity failure in `TestCli::testClip()` seems unrelated.

I wanted to check one thing: I had been assuming that the minimum supported Qt version is 5.9.5 (hence 0a1e77656505ec79492d371e9e4c8a3f9f632e74), as stated in [INSTALL.md](https://github.com/keepassxreboot/keepassxc/blob/develop/INSTALL.md#build-dependencies). The fact that #8829 recently got merged seems to contradict this. If we actually need to support Qt versions older than 5.9.5 then this PR might not work in its current form (but I'd also question why we want to support ancient Qt versions).

droidmonkey commented:
This PR is targeted towards 2.8.0 which I actually hope we'll upgrade to Qt 6.x as our minimum. I have zero desire to support any Qt less than 5.12 if we don't go to Qt 6. Nothing prevents people on older (imo broken) distros from continuing to use KeePassXC 2.7.x into the future.

c4rlo committed:
OpenSSHKeyGenDialog: fix use of deprecated setting

Use non-deprecated QComboBox::sizeAdjustPolicy setting

droidmonkey commented on src/gui/IconDownloader.cpp:
```diff
@@ -114,7 +103,7 @@ void IconDownloader::setUrl(const QString& entryUrl)
     QString secondLevelDomain;
     if (!hostIsIp) {
-        secondLevelDomain = getSecondLevelDomain(url);
+        secondLevelDomain = Tools::baseDomain(url);
```
This is actually wrong, we should just "chop up" a domain level (ie, my.example.com -> example.com) even if it means we end up with just a TLD (which will not resolve anyway). The current call here will always send us to the base domain potentially bypassing a level (level2.level1.domain.com -> domain.com, instead of level1.domain.com).

phoerious replied:
I think we should be using Mozilla's public suffix list here instead of naively chopping up domains. It requires the loading of another resource and some parsing logic for the list, but everything else is somewhat unsafe.

c4rlo replied:
@droidmonkey: My intent here was to keep the same behaviour. If you are saying the current behaviour is incorrect, I suggest fixing that in a separate PR. Or are you saying I am (accidentally) breaking the behaviour here?

@phoerious: The "clean" way to do this would be to use something like [libpsl](https://github.com/rockdaboot/libpsl). I just wanted to avoid adding another external dependency like that. And I feel that my current "unconventional" solution should be okay given that we have solid unit tests for this.

droidmonkey replied:
I am saying you are breaking the behavior.

c4rlo replied:
> we should just "chop up" a domain level (ie, my.example.com -> example.com) even if it means we end up with just a TLD (which will not resolve anyway)

Not sure I understand. Both before and after my change, we have `my.example.com` → `example.com` and `example.com` → `example.com`. This seems reasonable to me.

> The current call here will always send us to the base domain potentially bypassing a level (level2.level1.domain.com -> domain.com, instead of level1.domain.com).

That is also the old behaviour, even according to the source code comment for `getSecondLevelDomain()`.

The only case where I can think of a difference in behaviour is with an invalid URL like "`foo`", where `getSecondLevelDomain()` would presumably crash/throw in `parts.takeLast()` because `parts` is empty, whereas `Tools::baseDomain()` handles it fine as proven in the unit tests.

varjolintu replied:
I have made replacement for this using the Public Suffix List. I'm making a separate PR of it.

droidmonkey replied:
great, I really want to merge this one. I'll merge yours first though.

c4rlo replied:
Marking this resolved, since my latest push has deleted the relevant commit in favour of #9935.

droidmonkey commented on src/core/Tools.cpp:
```diff
@@ -300,6 +303,54 @@ namespace Tools
         return true;
     }
 
+    QString baseDomain(const QUrl& url)
+    {
+        QString domain = url.host();
```
Since this is inappropriately used in icon downloader, we should just move this to Browser code and out of tools to avoid the network dependency in unrelated modules.

c4rlo replied:
Marking resolved since this no longer applies.

c4rlo commented:
Hey, sorry about the radio silence here. I'm still up for getting this PR ready for merge, if the maintainers are interested. Now that #10267 is in progress to upgrade to Qt 6, I just want to check that it still makes sense to merge this first? I do think it still makes sense to do it that way – @varjolintu, thoughts?

varjolintu commented:
> Hey, sorry about the radio silence here. I'm still up for getting this PR ready for merge, if the maintainers are interested. Now that #10267 is in progress to upgrade to Qt 6, I just want to check that it still makes sense to merge this first? I do think it still makes sense to do it that way – @varjolintu, thoughts?

If there are no merge conflicts and everything looks good, then go ahead.

droidmonkey commented:
Yes still want to merge this first.

c4rlo commented:
Okay, great. I have resolved all merge conflicts and rebased this PR against `develop`. As part of that, I removed the commit that fixed how top-level domains are resolved, since #9935 has now taken care of that. ~~I have also added a couple extra commits that address some Botan deprecations.~~ _Edit:_ A couple of Botan deprecations are addressed separately in #10826; when that PR is combined with this one, a `-DWITH_DEV_BUILD=ON` build compiles successfully.

I have re-run the tests as stated in the original PR descriptions. Everything still works, except that when using an Ubuntu 18.04 container, there's now a few more CLI and GUI related test failures, but not sure how relevant those are [1]. Hopefully the GitHub checks that get run as part of this PR will show more relevant results.

[1] _Edit:_ I've compared these failures against the `develop` branch, and they all happen there too, so definitely nothing to worry about. For completeness, here they are:

```
FAIL!  : TestCli::testClip() '!errorOutput.contains("All clipping programs failed")' returned FALSE. ()
   Loc: [/keepassxc-src/tests/TestCli.cpp(653)]

FAIL!  : TestCli::testInvalidDbFiles() Compared values are not the same
   Actual   (QString(m_stderr->readAll()))                                                 : "Enter password to unlock /tmp/testcli.K17361: \nError while reading the database: Failed to read database file.\n"
   Expected (QObject::tr("Failed to open database file %1: not readable").arg(path) + "\n"): "Failed to open database file /tmp/testcli.K17361: not readable\n"
   Loc: [/keepassxc-src/tests/TestCli.cpp(2225)]
```

Again, this is in an Ubuntu 18.04 container; they don't occur on my Arch Linux machine. They look like artefacts of running the tests in a container, where some stuff doesn't work correctly.

droidmonkey commented on src/keys/drivers/YubiKey.h:
```diff
@@ -87,7 +94,11 @@ class YubiKey : public QObject
     bool m_initialized = false;
     QString m_error;
 
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+    static QRecursiveMutex s_interfaceMutex;
+#else
     static QMutex s_interfaceMutex;
+#endif
```
I don't even think we want a recursive mutex here anyway. This mutex is supposed to protect against multiple overlapping calls to find keys. These functions should not be recursively called in any normal context (and they aren't anyway).

TLDR - just change this to a standard QMutex

c4rlo replied:
Ok, done.

droidmonkey commented:
I'll merge this after 2.7.9 is bottled up, don't need any merge conflicts between branches.

c4rlo committed:
Make YubiKey::s_interfaceMutex non-recursive

It does not need to be a recursive mutex, as per
https://github.com/keepassxreboot/keepassxc/pull/7783#discussion_r1623274991.

c4rlo commented:
> I'll merge this after 2.7.9 is bottled up, don't need any merge conflicts between branches.

I've just rebased this again so there shouldn't be any merge conflicts (but don't mind if you want to defer merge after 2.7.9 anyway).
